c++11/include/opentracing/tracer.h

```diff
@@ -0,0 +1,198 @@
+#ifndef OPENTRACING_TRACER_H
+#define OPENTRACING_TRACER_H
+
+#include <opentracing/propagation.h>
+#include <opentracing/span.h>
+#include <opentracing/util.h>
+#include <initializer_list>
+#include <memory>
+#include <utility>
+#include <vector>
+
+namespace opentracing {
+// StartSpanOptions allows Tracer.StartSpan() callers  a mechanism to override
+// the start timestamp, specify Span References, and make a single Tag or
+// multiple Tags available at Span start time.
+//
+// StartSpan() callers should look at the StartSpanOption interface and
+// implementations available in this library.
+struct StartSpanOptions {
+  SystemTime start_system_timestamp;
+  SteadyTime start_steady_timestamp;
+  std::vector<std::pair<SpanReferenceType, const SpanContext*>> references;
+  std::vector<std::pair<std::string, Value>> tags;
+};
+
+// StartSpanOption instances (zero or more) may be passed to Tracer.StartSpan.
+class StartSpanOption {
+ public:
+  StartSpanOption(const StartSpanOption&) = delete;
+
+  virtual ~StartSpanOption() = default;
+
+  virtual void Apply(StartSpanOptions& options) const = 0;
+
+ protected:
+  StartSpanOption() = default;
+};
+
+// Tracer is a simple, thin interface for Span creation and SpanContext
+// propagation.
+class Tracer {
+ public:
+  virtual ~Tracer() = default;
+
+  // Create, start, and return a new Span with the given `operationName` and
+  // incorporate the given StartSpanOption `option_list`.
+  //
+  // A Span with no SpanReference options (e.g., opentracing::ChildOf() or
+  // opentracing::FollowsFrom()) becomes the root of its own trace.
+  //
+  // Examples:
+  //
+  //     opentracing::Tracer& tracer = ...
+  //
+  //     // The root-span case:
+  //     auto span = tracer.StartSpan("GetFeed")
+  //
+  //     // The vanilla child span case:
+  //     auto span = tracer.StartSpan(
+  //         "GetFeed",
+  //         {opentracing::ChildOf(parentSpan.context())})
+  //
+  //     // All the bells and whistles:
+  //     auto span = tracer.StartSpan(
+  //         "GetFeed",
+  //         {opentracing::ChildOf(parentSpan.context()),
+  //         opentracing::Tag{"user_agent", loggedReq.UserAgent},
+  //         opentracing::StartTimestamp(loggedReq.timestamp())})
+  //
+  std::unique_ptr<Span> StartSpan(
+      const std::string& operation_name,
+      std::initializer_list<option_wrapper<StartSpanOption>> option_list = {})
+      const {
+    StartSpanOptions options;
+    options.start_system_timestamp = SystemClock::now();
+    options.start_steady_timestamp = SteadyClock::now();
+    for (const auto& option : option_list) option.get().Apply(options);
+    return StartSpanWithOptions(operation_name, options);
+  }
```

Comment:
I was at Google for a hundred years and (because of their style guide) have very little experience programming with C++ exceptions. I'm not sure I have an opinion, so I recommend doing what you think is best.